Migrate klog to klog.InfoS in vendor/k8s.io/utils

hase1128 opened the pull request:
**What type of PR is this?**
 /kind cleanup

**What this PR does / why we need it**:

Ref:

kubernetes/enhancements#1602
https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging


**Explain about this PR**
Original klog call is `klog.Info(buffer.String())`

buffer.String() contains all trace information(trace name, start time, end time, etc) and single string message, so it needs to be decomposed into key and value.
However, there were the following problems.
A trace can contain multiple [steps](https://github.com/kubernetes/utils/tree/master/trace), 0 or more(the number of steps is not fixed.)
Each step contains trace information such as duration time.
I think that it is difficult to move this to InfoS with a fixed number of keys and values.

So, in this PR, we split one klog into three klog.InfoS
1. InfoS is called when trace starts
2. InfoS is called for each step (if there are multiple steps, InfoS is called same number of times.)
3. InfoS is called when the trace ends

The following is an example of log when the number of steps is 2.

Log output before migration
```
I0604 20:36:01.292061   32667 trace.go:116] Trace[2122920980]: "test-trace" test-key:100 (started: 2020-06-04 20:36:01.291233486 +0900 JST m=+67.466231521) (total time: 557.294μs):
Trace[2122920980]: [441.259μs] [441.259μs] step1-trace
Trace[2122920980]: [443.542μs] [2.283μs] step2-trace key-B:200,test-key:100,key-A:testtest
Trace[2122920980]: [557.294μs] [113.752μs] END
```

After migration
```
I0604 20:57:25.449051   16582 trace.go:94] "Trace" tracenum=1438512885 name="test-trace" keyvalues="test-key:100" starttime="2020-06-04 20:57:25.448668419 +0900 JST m=+67.710287384" totaltime="307.95μs"
I0604 20:57:25.449107   16582 trace.go:103] "Trace" tracenum=1438512885 steptime="245.901μs" stepduration="245.901μs" stepmsg="step1-trace" keyvalues=""
I0604 20:57:25.449198   16582 trace.go:103] "Trace" tracenum=1438512885 steptime="248.113μs" stepduration="2.212μs" stepmsg="step2-trace" keyvalues="key-B:200,test-key:100,key-A:testtest"
I0604 20:57:25.449268   16582 trace.go:109] "Trace end" tracenum=1438512885 endtime="307.95μs" stepduration="59.837μs"
```

In the log before migration, new line is inserted in each above 1 to 3 (Trace start/each step/Trace end), so the format is similar to after migration( eventhough there is multipule InfoS).

Is this idea(multiple InfoS) accepted?
Please let me know if we need to change it.

hase1128 committed:
Migrate klog to klog.InfoS in vendor/k8s.io/utils

hase1128 commented:
/assign @sttts
/assign @serathius

serathius commented on vendor/k8s.io/utils/trace/trace.go:
```diff
@@ -88,32 +88,27 @@ func (t *Trace) logWithStepThreshold(stepThreshold time.Duration) {
-		buffer.WriteString(" ")
 	}
-	buffer.WriteString(fmt.Sprintf("(started: %v) (total time: %v):\n", t.startTime, totalTime))
+	klog.InfoS("Trace", "tracenum", tracenum, "name", t.name, "keyvalues", buffer.String(), "starttime", t.startTime, "totaltime", totalTime)
```
```suggestion
	klog.InfoS("Trace start", "tracenum", tracenum, "name", t.name, "keyvalues", buffer.String(), "starttime", t.startTime, "totaltime", totalTime)
```

serathius commented on vendor/k8s.io/utils/trace/trace.go:
```diff
@@ -88,32 +88,27 @@ func (t *Trace) logWithStepThreshold(stepThreshold time.Duration) {
 				writeFields(&buffer, step.fields)
 			}
-			buffer.WriteString("\n")
+			klog.InfoS("Trace", "tracenum", tracenum, "steptime", step.stepTime.Sub(t.startTime), "stepduration", stepDuration, "stepmsg", step.msg, "keyvalues", buffer.String())
```
```suggestion
			klog.InfoS("Trace step", "traceNum", tracenum, "stepTime", step.stepTime.Sub(t.startTime), "stepDuration", stepDuration, "stepMsg", step.msg, "keyvalues", buffer.String())
```

serathius commented on vendor/k8s.io/utils/trace/trace.go:
```diff
@@ -88,32 +88,27 @@ func (t *Trace) logWithStepThreshold(stepThreshold time.Duration) {
 	stepDuration := endTime.Sub(lastStepTime)
 	if stepThreshold == 0 || stepDuration > stepThreshold || klog.V(4).Enabled() {
-		buffer.WriteString(fmt.Sprintf("Trace[%d]: [%v] [%v] END\n", tracenum, endTime.Sub(t.startTime), stepDuration))
+		klog.InfoS("Trace end", "tracenum", tracenum, "endtime", endTime.Sub(t.startTime), "stepduration", stepDuration)
```
```suggestion
		klog.InfoS("Trace end", "traceNum", tracenum, "endTime", endTime.Sub(t.startTime), "stepDuration", stepDuration)
```

serathius commented on vendor/k8s.io/utils/trace/trace.go:
```diff
@@ -88,32 +88,27 @@ func (t *Trace) logWithStepThreshold(stepThreshold time.Duration) {
+			buffer.Reset()
 			if len(step.fields) > 0 {
-				buffer.WriteString(" ")
 				writeFields(&buffer, step.fields)
```
Please flatten fields and pass them into klog.InfoS call

```
func logFields(fields []Field) []interface {
  result := make([]interface, 0, 2*len(fields)
  for _, field := range fields { 
    result = append(result, f.Key, f.Value)  
  }
  return result
}

klog.InfoS("Trace", "tracenum", tracenum, "steptime", step.stepTime.Sub(t.startTime), "stepduration", stepDuration, "stepmsg", step.msg, logFields(step.fields)...)
```

hase1128 replied:
In this case, I see the following error.
```
have (string, string, int32, string, time.Duration, string, time.Duration, string, string, []interface {}...)
want (string, ...interface {})
```

Can we use slice of interface{} after multiple key and values?
or do other values (startTime, traceNum, etc) also need to be included in the same slice as fields, making it **a single slice**?

serathius replied:
We need to put whole log into single slice (forgot about this)
```
klog.InfoS("Trace step", logStepArgs(step)...)

func logStepArgs(traceNum  int, step Step, stepDuration time.Duration) []interface {
  result := make([]interface, 0, 2*len(step.fields) + 8
  result = append(result, "traceNum", tracenum, "stepTime", step.stepTime.Sub(t.startTime), "stepDuration", stepDuration, "stepMsg", step.msg)
  for _, field := range fields { 
    result = append(result, f.Key, f.Value)  
  }
  return result
}
```

hase1128 replied:
I see, thanks

serathius commented:
Idea looks good.

Proposed improvements:
* Name of fields should use lowerCamelCase. 
* Flatten keyvalues using `logFields(step.fields)...`

yuzhiquan commented:
/release-note-none

hase1128 committed:
Add functions for Filed slice

hase1128 commented:
> Idea looks good.
> 
> Proposed improvements:
> 
> * Name of fields should use lowerCamelCase.
> * Flatten keyvalues using `logFields(step.fields)...`

I addressd the above, and I removed unnecessary process(buffer writing) any more.
I also added function for root trace(This is similar to logStepArgs), because not only step but also root trace have Field slice.

Log output is here.

before
```
I0604 20:36:01.292061   32667 trace.go:116] Trace[2122920980]: "test-trace" test-key:100 (started: 2020-06-04 20:36:01.291233486 +0900 JST m=+67.466231521) (total time: 557.294μs):
Trace[2122920980]: [441.259μs] [441.259μs] step1-trace
Trace[2122920980]: [443.542μs] [2.283μs] step2-trace key-B:200,test-key:100,key-A:testtest
Trace[2122920980]: [557.294μs] [113.752μs] END
```

after
```
I0605 18:26:45.591104   28913 trace.go:109] "Trace start" traceNum=1335657781 traceName="test-trace" test-key=100 startTime="2020-06-05 18:26:45.59063556 +0900 JST m=+67.804229827" totalTime="384.704μs"
I0605 18:26:45.591145   28913 trace.go:114] "Trace step" traceNum=1335657781 stepTime="343.928μs" stepDuration="343.928μs" stepMsg="step1-trace"
I0605 18:26:45.591172   28913 trace.go:114] "Trace step" traceNum=1335657781 stepTime="345.644μs" stepDuration="1.716μs" stepMsg="step2-trace" key-B=200 test-key=100 key-A="testtest"
I0605 18:26:45.591194   28913 trace.go:120] "Trace end" traceNum=1335657781 endTime="384.704μs" stepDuration="39.06μs"
```

It is correctly output when there is no Field, when there is one, or when there are multiple Fields.

yuzhiquan commented:
PLZ squash commit.
/ok-to-test

serathius commented on vendor/k8s.io/utils/trace/trace.go:
```diff
@@ -44,6 +44,25 @@ func writeFields(b *bytes.Buffer, l []Field) {
+	for _, field := range trace.fields {
+		result = append(result, field.Key, field.Value)
+	}
+	result = append(result, "startTime", trace.startTime, "totalTime", totalTime)
```
Why split append into 2 calls?

hase1128 replied:
Fields[] is outputed in the middle in original trace log message:
(tracenum, t.name, Field[], starttime, totaltime)
```
	buffer.WriteString(fmt.Sprintf("Trace[%d]: %q ", tracenum, t.name))
	if len(t.fields) > 0 {
		writeFields(&buffer, t.fields)
		buffer.WriteString(" ")
	}
	buffer.WriteString(fmt.Sprintf("(started: %v) (total time: %v):\n", t.startTime, totalTime))
```

So, I splited in order to keep order of element of trace as the original.
If Field[] is outputtd the end of original log message, we don't need to split.

hase1128 replied:
Is there any modification about this?

serathius commented on vendor/k8s.io/utils/trace/trace.go:
```diff
@@ -83,37 +102,24 @@ func (t *Trace) Log() {
-		buffer.WriteString(" ")
-	}
-	buffer.WriteString(fmt.Sprintf("(started: %v) (total time: %v):\n", t.startTime, totalTime))
+	klog.InfoS("Trace start", logTraceArgs(tracenum, t, totalTime)...)
```
Think to consider: depending on klog verbosity we will get two outputs

if `klog.V(4).Enabled()`:
```
Trace start ...
Trace step ...
Trace stop ...
```
else
```
Trace start
```

It could be confusing for users that there is a `Trace start` message without `Trace end`.
I see two options
* Change `Trace start` to `Trace`
* depending on `klog.V(4).Enabled()` change message

serathius replied:
I think option 1 would make more sense.

hase1128 replied:
I think option1 is good. I modify this.

hase1128 committed:
Modify trace message to address klog verbosity

hase1128 commented:
Updated log message:

before migration
```
I0604 20:36:01.292061   32667 trace.go:116] Trace[2122920980]: "test-trace" test-key:100 (started: 2020-06-04 20:36:01.291233486 +0900 JST m=+67.466231521) (total time: 557.294μs):
Trace[2122920980]: [441.259μs] [441.259μs] step1-trace
Trace[2122920980]: [443.542μs] [2.283μs] step2-trace key-B:200,test-key:100,key-A:testtest
Trace[2122920980]: [557.294μs] [113.752μs] END
```

after
```
I0608 14:48:44.534733    7968 trace.go:109] "Trace" traceNum=1881226924 traceName="test-trace" test-key=100 startTime="2020-06-08 14:48:44.534150253 +0900 JST m=+67.503938280" totalTime="500.061μs"
I0608 14:48:44.534802    7968 trace.go:114] "Trace" traceNum=1881226924 stepTime="352.674μs" stepDuration="352.674μs" stepMsg="step1-trace"
I0608 14:48:44.534845    7968 trace.go:114] "Trace" traceNum=1881226924 stepTime="354.178μs" stepDuration="1.504μs" stepMsg="step2-trace" key-B=200 test-key=100 key-A="testtest"
I0608 14:48:44.534885    7968 trace.go:120] "Trace" traceNum=1881226924 endTime="500.061μs" stepDuration="145.883μs"
```

hase1128 commented:
/retest

hase1128 commented:
/retest

wojtek-t commented:
This PR is incorrect. 
vendor/ directory can't be changed itself.
You should update that in the original repo (k8s.io/utils) [send a PR against that repo] and then bump depenendency here.

@hase1128 @serathius ^^

serathius commented:
Yes, @wojtek-t is right. Sorry for not noticing this earlier. 
@hase1128 Can you file a PR to https://github.com/kubernetes/utils  ?

hase1128 commented:
@serathius 
I issued PR to kubernetes/utils. Please check.
https://github.com/kubernetes/utils/pull/171

serathius commented:
/remove-lifecycle stale

ehashman commented:
I will close this PR; as @wojtek-t said above, we should not be trying to update the vendor/ directory directly.

Please open a new PR once you've migrated the components to bump the dependencies :)

/close
